### ISSUE TITLE
state.solver.simplify messes up with preconstrainer

### ISSUE THREAD

**Issue by Kyle-Kyle** [angr/angr#1460]
**Describe the bug**
Please include a clear and concise description of what the bug is.
When `preconstrainer` is used and a `state.solver.simplify` is called.
Collected constraints will be deleted because preconstrained constraints are better.

**To Reproduce**
Please include *both a script to reproduce the crash, and attach the binary used, if possible*

~~~
import angr
import tracer
from angr.state_plugins.preconstrainer import SimStatePreconstrainer
from angr.state_plugins.posix import SimSystemPosix
from angr.storage.file import SimFileStream

import logging
logging.getLogger('angr.exploration_techniques').setLevel('DEBUG')

path = '/home/kylebot/src/angr-dev/binaries/tests/cgc/cfe_CADET_00003'
inp = b'A'*0x80

proj = angr.Project(path)

input_file = SimFileStream(name='stdin', ident='aeg_stdin')
state = proj.factory.full_init_state(stdin=input_file)
state.preconstrainer.preconstrain_file(inp, input_file, set_length=True)

simgr = proj.factory.simgr(state)
runner = tracer.QEMURunner(path, inp)
tracer_tech = angr.exploration_techniques.Tracer(trace=runner.trace, crash_addr=runner.crash_addr)
simgr.use_technique(tracer_tech)

simgr.run()

found = simgr.traced[0]

print('preconstraint:', len(found.preconstrainer.preconstraints))
print('before simplification:', len(found.solver.constraints))
found.solver.simplify()
print('after simplification:', len(found.solver.constraints))
found.preconstrainer.remove_preconstraints()
print('after remove preconstraints:', len(found.solver.constraints))
~~~

**Comment by Kyle-Kyle** [comment]
output:
~~~
preconstraint: 128
before simplification: 326
after simplification: 128
after remove preconstraints: 0
~~~

**Comment by Kyle-Kyle** [comment]
My proposal solution was adding a check in `state.solver.simplify()` to see whether preconstrainer is in use.
But this design could be very bad. `simplify` should just simplify all constraints.
What do you think about giving out a warning?

**Comment by rhelmot** [comment]
perhaps the best solution would be reimplemnt the preconstrainer as a subclass of SimSolver. That would give it very nice control over how its special values are treated.

**Comment by Kyle-Kyle** [comment]
I'm not talking the implementation. It's about the design.
I still think `solver.simplify` should just simplify all `solver.constraints`. Either we separate preconstraints from collected constraints(so it's never in `solver.constraints`)  or we let user know that they are doing something bad.

**Comment by Kyle-Kyle** [comment]
That's only my personal opinion.

**Comment by rhelmot** [comment]
My preference is that simplification would not affect preconstraints.
However that's not really based on anything so if you have your own
opinions feel free to implement them

**Comment by Kyle-Kyle** [comment]
should be fixed by #3250
